bigtable-hbase-mirroring-client-1.x-parent/bigtable-hbase-mirroring-client-1.x/src/test/java/com/google/cloud/bigtable/mirroring/hbase1_x/TestMirroringTable.java

```diff
@@ -0,0 +1,1411 @@
+        put1,
+        Result.create(new Cell[0]),
+        put2,
+        Result.create(new Cell[0]),
+        get1,
+        Result.create(new Cell[0]));
+
+    Object[] results = new Object[3];
+    mirroringTable.batch(request, results);
+    executorServiceRule.waitForExecutor();
+
+    verify(primaryTable, times(1)).batch(request, results);
+    verify(secondaryTable, never()).batch(ArgumentMatchers.<Row>anyList(), any(Object[].class));
+    verify(secondaryWriteErrorConsumer, times(1))
+        .consume(eq(HBaseOperation.BATCH), eq(ImmutableList.of(put1, put2)), any(Throwable.class));
+  }
+
+  @Test
+  public void testBatchWithAppendsAndIncrements() throws IOException, InterruptedException {
+    Increment increment = new Increment("i".getBytes());
+    increment.addColumn("f".getBytes(), "q".getBytes(), 1);
+
+    Append append = new Append("a".getBytes());
+    append.add("f".getBytes(), "q".getBytes(), "v".getBytes());
+
+    Put put = createPut("p", "f", "q", "v");
+    Get get = createGet("g");
+    List<? extends Row> operations = Arrays.asList(increment, append, put, get);
+    mockBatch(
+        primaryTable,
+        increment,
+        createResult("i", "f", "q", 1, "1"),
+        append,
+        createResult("a", "f", "q", 2, "2"),
+        put,
+        new Result(),
+        get,
+        createResult("g", "f", "q", 3, "3"));
+    Object[] results = new Object[operations.size()];
+
+    List<? extends Row> expectedSecondaryOperations =
+        Arrays.asList(
+            createPut("i", "f", "q", 1, "1"),
+            createPut("a", "f", "q", 2, "2"),
+            createPut("p", "f", "q", "v"),
+            createGet("g"));
+    mirroringTable.batch(operations, results);
+    executorServiceRule.waitForExecutor();
+    verify(primaryTable, times(1)).batch(operations, results);
+    ArgumentCaptor<List<? extends Row>> argumentCaptor = ArgumentCaptor.forClass(List.class);
+    verify(secondaryTable, times(1)).batch(argumentCaptor.capture(), any(Object[].class));
+
+    assertPutsAreEqual(
+        (Put) argumentCaptor.getValue().get(0), (Put) expectedSecondaryOperations.get(0));
+    assertPutsAreEqual(
+        (Put) argumentCaptor.getValue().get(1), (Put) expectedSecondaryOperations.get(1));
+    assertPutsAreEqual(
+        (Put) argumentCaptor.getValue().get(2), (Put) expectedSecondaryOperations.get(2));
+    assertThat(argumentCaptor.getValue().get(3)).isEqualTo(expectedSecondaryOperations.get(3));
+  }
+
+  @Test
+  public void testConcurrentWritesAreFlowControlledBeforePrimaryAction()
+      throws IOException, InterruptedException {
+    this.mirroringTable =
+        spy(
+            new MirroringTable(
+                primaryTable,
+                secondaryTable,
+                this.executorServiceRule.executorService,
+                mismatchDetector,
+                flowController,
+                secondaryWriteErrorConsumer,
+                new ReadSampler(100),
+                true,
```

Comment:
Add a comment explaining htat this sets the concurrent operations

Reply:
Done.